src/main/java/com/zendesk/maxwell/producer/MaxwellPubsubProducer.java

```diff
@@ -0,0 +1,225 @@
+  static final Logger LOGGER = LoggerFactory.getLogger(MaxwellPubsubProducer.class);
+
+  private final String projectId;
+  private Publisher pubsub;
+  private final TopicName topic;
+  private final TopicName ddlTopic;
+  private Publisher ddlPubsub;
+  private final Timer metricsTimer;
+  private final ArrayBlockingQueue<RowMap> queue;
+  private Thread thread;
+  private StoppableTaskState taskState;
+
+  public MaxwellPubsubProducerWorker(MaxwellContext context, Properties pubsubProperties, String pubsubProjectId, String pubsubTopic, ArrayBlockingQueue<RowMap> queue) {
+    super(context);
+
+    if ( pubsubTopic == null ) {
+      this.topic = TopicName.create(pubsubProjectId, "maxwell");
+    } else {
+      this.topic = TopicName.create(pubsubProjectId, pubsubTopic);
+    }
+
+    this.projectId = pubsubProjectId;
+    this.ddlTopic = TopicName.create(pubsubProjectId, context.getConfig().ddlPubsubTopic);
+
+    Metrics metrics = context.getMetrics();
+    this.metricsTimer = metrics.getRegistry().timer(metrics.metricName("message", "publish", "time"));
+
+    this.queue = queue;
+    this.taskState = new StoppableTaskState("MaxwellPubsubProducerWorker");
+  }
+
+  public void setPubsub() throws IOException {
+    this.pubsub = Publisher.defaultBuilder(this.topic).build();
+  }
+
+  public void setDDLPubsub() throws IOException {
+    this.ddlPubsub = Publisher.defaultBuilder(this.ddlTopic).build();
+  }
+
+  @Override
+  public void run() {
+    this.thread = Thread.currentThread();
+    while ( true ) {
+      try {
+        RowMap row = queue.take();
+        if ( !taskState.isRunning() ) {
+          taskState.stopped();
+          return;
+        }
+        this.push(row);
+      } catch ( Exception e ) {
+        taskState.stopped();
+        context.terminate(e);
+        return;
+      }
+    }
+  }
+
+  @Override
+  public void sendAsync(RowMap r, AbstractAsyncProducer.CallbackCompleter cc) throws Exception {
+    String message = r.toJSON(outputConfig);
+    ByteString data = ByteString.copyFromUtf8(message);
+    PubsubMessage pubsubMessage = PubsubMessage.newBuilder().setData(data).build();
+
+    if ( r instanceof DDLMap ) {
+      ddlPubsub.publish(pubsubMessage);
+    } else {
+      pubsub.publish(pubsubMessage);
+    }
+  }
+
+  @Override
+  public void requestStop() {
+    taskState.requestStop();
+    try {
+      pubsub.shutdown();
+    } catch (Exception e) {
+      LOGGER.error("Failed to shutdown Pubsub publisher: " + e.getLocalizedMessage());
+      LOGGER.error("    projectId: " + projectId);
+      LOGGER.error("    topic: " + topic.getTopic());
+    }
+
+    if ( ddlPubsub != null ) {
+      try {
+        ddlPubsub.shutdown();
+      } catch (Exception e) {
+        LOGGER.error("Failed to shutdown Pubsub publisher: " + e.getLocalizedMessage());
+        LOGGER.error("    projectId: " + projectId);
+        LOGGER.error("    topic: " + ddlTopic.getTopic());
+      }
+    }
+  }
+
```

Comment:
`requestStop` will be allowed to throw an exception once #742 is merged - are you happy if we later take out these catches and just propagate the error?

Reply:
Yeah! Also just noticed you've since merged #742, so removed the catches now.